### ISSUE TITLE
Cluster Autoscaler - support additional flag customization

### ISSUE THREAD

**Issue by antoinne85** [Azure/AKS#1055]
**What happened**:
The autoscaler scaled my cluster up by 1 instance.

**What you expected to happen**:
The autoscaler would scale my cluster up enough to schedule the unschedulable pods.

**How to reproduce it (as minimally and precisely as possible)**:
Create a cluster with the autoscaler enabled. Drop a deployment onto the cluster that has many replicas and resource requests such that several pods are unschedulable.

**Anything else we need to know?**:
The [documentation](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca) for the cluster autoscaler suggests that there are several parameters you can tweak to control its behavior, but I don't actually _see_ the autoscaler deployed onto the cluster (even though I see the VMSS being scaled up/down) and I can't find any documentation on affecting the autoscaler's behavior in AKS.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.8
- Size of cluster (how many worker nodes are in the cluster?) 3 nodes
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.) microservices, batch processes, basically doing nothing since I'm just experimenting with the autoscaler

**Comment by jnoller** [comment]
@feiskyer ^

**Comment by feiskyer** [comment]
@antoinne85 The cluster autoscaler is managed by AKS, hence you won't see any pods running on your nodes. May I ask what parameters are you trying to tune?

**Comment by antoinne85** [comment]
@feiskyer Thanks for getting back to me! To be completely honest, I'm not 100% super which parameters, _in particular_, I'd be looking to tune, but I can describe to you the situation that made me start looking into what my _options_ are.

### TL;DR Version
I don't know _what_ parameters I need/want to tune. What I _do_ know is that I pushed a Deployment onto the cluster that would have required several nodes to be added to the cluster in order to schedule all of the Pods, but only node was added. 

### The Long Version
We deploy onto the cluster using Helm. Our CI system build, tests, containerizes and finally creates a helm chart that describes the release and its dependencies, including a script that gets run to actually update the version of the chart that's deployed onto the cluster (deployments are handled by Octopus Deploy, not that it's super-relevant, but it's not as if there's a human doing the deployments).

As you might expect, the critical part of the deployment, when done is way, is the call to `helm upgrade`. In our case, we pass the `--wait` flag to instruct Helm to wait for everything to come online before the release is marked as successful.

Now that you've got the background--

We stood up a fresh cluster with autoscaling enabled and started deploying our applications onto it with the intent of testing out our workloads with autoscaling. One of those applications included a `Deployment` with an instance count of 20 or so. Naturally, the cluster didn't have enough resources (at only 3 nodes) to run this application with 20 instances, so the autoscaler kicked in and added a node--_but just one node_. As a result, many of the instances never came online and so Helm hit its timeout and declared the release as failed.

While we can obviously increase the timeout for Helm, that comes at the cost of timely deployments. As with everything else, its a tradeoff. In the typical case, where the cluster has the capacity, a deployment will take a minute or two. In the worst case, I can't even predict how long it would have taken to deploy this workload because I don't have access to things like:

* Minimum time between scaling actions
* Controls around the min/max instance change for a single scaling action

It's not very ideal to, in a scenario like this, know immediately that you need, just as an example, 5 more nodes to schedule all the pods, and then add only a single node at a time, waiting for the cooldown period between scaling actions between each node being added. Just some fuzzy math--if we assume:
* Time between scaling actions is 1 minute and is counted from the time the previous action finished.
* Time to bring a node online is 2 minutes.

If I deploy and can see that I need 3 more nodes to schedule all the Pods it'll be:
* 2 minutes to bring on New Node 1
* 1 minute cooldown
* 2 minutes to bring on New Node 2
* 1 minute cooldown
* 2 minutes to bring on New Node 3

For a total of 8 minutes, when I would have expected all three nodes to come online together and only be waiting around 2 minutes total.

To be fair, I'm just making up numbers and there are some assumptions about behavior baked in here (like the cooldown between scaling actions), but those assumptions are based on the behavior I _saw_. In my case the helm timeout was long enough for the _one new node_ to come online, but then I didn't see another scaling action in progress when I looked at the VMSS in the portal (you probably already know this, but when a scaling action is in progress it tells you in the UI).

Again--thanks for taking the time to join in here, I'm happy to provide any additional detail or re-run the trial if it help at all.

**Comment by antoinne85** [comment]
I should probably add, since I didn't _really_ answer your question--

If I were to _guess_ at the specific parameters that I think I would find useful, they'd be:
* Some way to overprovision (e.g. the equivalent of 2 nodes worth of resources must be free on the cluster at all times)
* Some way to control the minimum time between scaling actions (probably different values for up/down)
* Have it scale up by the number of nodes it _need_ to schedule _all_ unscheduled Pods (or, in the absence of that, some way to set the "scale step" size to have it always scale up by some number of instances, with the expectation that it _might_ be too much and it'll just wind up scaling back down in a bit).

**Comment by feiskyer** [comment]
For overprovisioning, please refer the detailed guide on CA [FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-configure-overprovisioning-with-cluster-autoscaler).

For the various options (e.g. intervals), AKS has set default values for them and they're not exposed for customers. cc @jluk for evaluating whether they could be exposed.

**Comment by jluk** [comment]
Agree with @feiskyer we can look into exposing additional flags, have heard this from other folks as well. Let us take a look while we work on improving CA on AKS to GA quality.

**Comment by antoinne85** [comment]
@jluk Thanks!

Do I need to submit something on UserVoice for the purposes of tracking any development or communication around this topic? I guess, basically what I'm asking, is that I'd prefer not to annoy you every 3 weeks or so with "any update?" and instead just have a clear understanding of how I should expect further communication to arrive. And if you've said nothing I'll assume there's nothing to say yet.

**Comment by jluk** [comment]
@antoinne85 this issue should be good! I've assigned myself and a few labels on this, to set expectation I'm hoping to have an update in 4-6 weeks as we're pushing through some existing backlog work but I am onboard with this ask. Feel free to chime back in on this with new related ask / scenario specifics to help us drive it.

**Comment by jluk** [comment]
Providing update here as more improvements are coming in this month for this feature - adding @palma21 who will be participating in this improvement.

**Comment by jluk** [comment]
Additional flags have been shipped - doc below capturing all the new configurations and defaults. 

Closing this issue as a result, but feel free to continue leaving feedback. Thanks!

https://docs.microsoft.com/azure/aks/cluster-autoscaler#using-the-autoscaler-profile

**Comment by davelosert** [comment]
May I ask why you only enabled those 8 specific flags to be updated?

In our case, we would like to have an evenly balanced cluster across several nodepools aka groups - meaning we would need to set `--balance-similar-node-groups` flag to true.

Are there any plans to extend the number of adjustable flags?

**Comment by marwanad** [comment]
@davelosert Thanks for raising this up. There were plans to default this value to true before the work on flags started. We'll add this to the list of parameters.

The plan was to start with a small number of flags (given that CAS has 20+ flags) and keep expanding on those based on feedback.

**Comment by davelosert** [comment]
@marwanad : Thanks for the quick reply. That sounds like good news 😄 

Shall I go ahread and raise a new issue for that topic? Or are you tracking this already internally? And is there any estimate onto when this might be added?

**Comment by jluk** [comment]
@davelosert please open a new issue as a request with the specifics you need, thanks for the feedback!
